### ISSUE TITLE
How is the chat example P2P

### ISSUE THREAD

**Issue by kishansagathiya** [libp2p/go-libp2p#319]
https://github.com/libp2p/go-libp2p/tree/master/examples/chat

It asks B to create a listener(server) and A to connect(client). I will still face the same problem of NAT Traversal. How is this P2P? Seems more like server client to me.

**Comment by upperwal** [comment]
> Are you saying that it will work without me doing any port forwarding on my router?

Yes it's possible. Protocols like UPnP and PMP does this for you. (although UPnP should be enabled in your router)

**Comment by upperwal** [comment]
@Harrywang55666 

It's as simple as adding ```libp2p.NATPortMap()``` to the following code.
https://github.com/libp2p/go-libp2p/blob/9e5ef7a4e6ced5eb5eb9982b2a0541b331ba5387/examples/chat/chat.go#L169-L173

like 

```go
host, err := libp2p.New( 
	context.Background(), 
	libp2p.ListenAddrs(sourceMultiAddr), 
	libp2p.Identity(prvKey), 
	libp2p.NATPortMap(),
 )
```
this will open a port in your network's firewall using UPnP. Ref [docs](https://godoc.org/github.com/libp2p/go-libp2p#NATPortMap) for more.

But you should keep two things in mind.
1. Libp2p by default uses Hole Punching (another NAT traversal technique) via ```SO_REUSEPORT``` option. So you can listen on the same port you are dialling from.
2. This example needs you to manually specify a destination peer to establish a connection. If you want to connect automatically, use some bootstrapping technique (like DHT). I tried this w/o ```libp2p.NATPortMap()``` and it worked.

Let me know if this does not clear your doubt.

**Comment by bigs** [comment]
i've got a PR up to update the chat example to reflect more modern usage of the library over here #332

**Comment by hellogeraldblah** [comment]
Hello, I have read through your discussion above and I'm facing some connectivity issues. I used the chat.go example from https://github.com/libp2p/go-libp2p/tree/master/examples/chat, and added in the libp2p.NATPortMap() option into libp2p.New() as well. 

However, when I run node A, node B (different network) attempts to connect to node A and receives a timeout error. I have also checked that my router has UPnP enabled.. 

If I understand it right, if node A (network A) and node B (network B) wishes to connect to each other directly, they can utilise libp2p.NATPortMap() which uses the hole punch concept.

May I know where I might have gone wrong or misunderstood certain concepts?
